### DISCUSSION TITLE
JSON object scalar

### DISCUSSION THREAD

**Discussion by nitsky** [graphql-rust/juniper#1076] [Q&A]
Hi, I would like to create a custom scalar that serializes as a JSON object. @tyranron is this feasible with the v0.15 API, or are #1072 and #975 necessary for this?

I am hoping to be able to implement a GraphQL server like this:

```graphql
scalar Point

type Query {
  location: Point
}
```

```rust
#[derive(serde::Serialize)]
struct Point {
  x: f32,
  y: f32,
}
```

**Comment by tyranron** [comment]
@nitsky yes, for having fields in GraphQL scalar like this we need to finish https://github.com/graphql-rust/juniper/pull/1072 and https://github.com/graphql-rust/juniper/pull/975 first, which is definitely 0.16 and onwards.

But why not use `GraphQLInputObject` for this?

**Reply by nitsky** [reply]
@tyranron I would like to use these types for output as well as input, and the actual struct I have has serialize and deserialize impls with serde that would be difficult to express again with juniper. Support for serde would be ideal here. The graphql-client crate in this org already supports this. In the meantime, would you recommend I serialize JSON to a string, and include that in graphql?

**Reply by ilslv** [reply]
@nitsky 

> Support for serde would be ideal here. The graphql-client crate in this org already supports this.

GraphQL client doesn't validate queries, so there is no need to have information of concrete field types. For GraphQL server we have to know a lot more about type to validate queries, so using only `serde` isn't really possible. 

> would you recommend I serialize JSON to a string, and include that in graphql?

I would recommend having 2 structs: one for input object and another for output

```rust
#[derive(serde::Serialize, GraphQLObject)]
struct Point {
  x: f32,
  y: f32,
}

#[derive(GraphQLInputObject)]
struct PointInput {
  x: f32,
  y: f32,
}

impl From<PointInput> for Point { ... }
```

@tyranron maybe we should try allowing deriving `GraphQLObject` and `GraphQLInputObject` for convenience? `async-graphql` [already does that](https://github.com/async-graphql/async-graphql/pull/697). Though I'm not sure this is possible without more changes to core traits as pointed out [here](https://github.com/graphql-rust/juniper/issues/203)

**Reply by tyranron** [reply]
@ilslv yes, nice idea. With granular core traits in #1072 it would be possible.

**Reply by nitsky** [reply]
This would work in the simple case of the `Point` example I gave, but the data structures I am dealing with have a complex serialization defined with serde including multiple levels of `struct`s and `enum`s, and I don't want to allow users of my API to attempt to query a subset of the fields, it's either all or nothing, so a scalar is the most appropriate. Is serializing JSON to a string my only option until #1072 and #975 land?

**Reply by ilslv** [reply]
@nitsky 

> I don't want to allow users of my API to attempt to query a subset of the fields, it's either all or nothing, so a scalar is the most appropriate.

If you want to just give users a lot of data without querying subsets of it, custom scalar is your best option. I don't see how support for `serde_json` would help here. This integration is more for dynamic schema which isn't defined strictly with Rust types. 
`0.16` also will land [more ergonomic scalar macros](https://graphql-rust.github.io/juniper/master/types/scalars.html#graphqlwith--path-attribute).

**Reply by nitsky** [reply]
I guess my question is, can my custom scalar be a complex nested JSON value, which I serialize with serde? Does that make sense?

**Reply by tyranron** [reply]
@nitsky 

> Is serializing JSON to a string my only option until https://github.com/graphql-rust/juniper/pull/1072 and https://github.com/graphql-rust/juniper/pull/975 land?

Yes.

**Reply by tyranron** [reply]
@nitsky 

> I guess my question is, can my custom scalar be a complex nested JSON value, which I serialize with serde? 

Something like this should work now:
```rust
#[derive(Deserialize, Serialize, GraphQLScalar)]
#[graphql(parse_token(String))]
struct MyCustomStruct {
    // a lot of fields...
}

impl MyCustomStruct {
    fn from_input<S: ScalarValue>(
        input: &InputValue<S>,
    ) -> Result<Self, String> {
        serde_json::from_str(input.as_string_value()).map_err(|e| e.to_string())
    }

    fn to_output<S: ScalarValue>(&self) -> Value<S> {
        serde_json::to_string(self).unwrap().into()
    }
}
```
